gputils/doc/gpvo.1
.TH gpvo 1 "(c) 2002, 2003 James Bowman, Craig Franklin"
.SH NAME
gpvo - GNU PIC object file Viewer
.SH SYNOPSIS
.B gpvo [options] file
.SH WARNING
The information in this man page is an extract from the full documentation of
gputils and is limited to the meaning of the options.  For complete and 
current documentation, refer to "gputils.ps" located in the gputils docs 
directory.
.SH DESCRIPTION
.B gpvo
is an COFF object file viewer for Microchip (TM) PIC (TM) micro-controllers.  
COFF files are relocatable objects output from gpasm.
.B gpvo
is part of gputils.  Check the gputils(1) manpage for details on other GNU 
PIC utilities.
.SH OPTIONS
.TP
.B -b, --binary
Display binary data.
.TP
.B -f, --file
Display file header.
.TP
.B -h, --help
Show the usage message. 
.TP
.B -o, --optional 
Display optional header.
.TP
.B -r, --string
Display the string table.
.TP
.B -s, --section 
Display sections.
.TP
.B -t, --symbol
Display symbol table.
.TP
.B -v, --version
Show the version.
.SH SEE ALSO
gputils(1)
.SH AUTHOR
Craig Franklin <craigfranklin@attbi.com>
.SH COPYRIGHT
Copyright (C) 2002, 2003 James Bowman, Craig Franklin

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


gputils/ChangeLog
2014-01-25  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p12f1612.inc, header/p12lf1612.inc, header/p16f1613.inc,
    header/p16lf1613.inc, header/p16f1xxx.inc, libgputils/gpcfg-table.c,
    libgputils/gpprocessor.c, lkr/Makefile.am, lkr/Makefile.in,
    lkr/12f1612_g.lkr, lkr/16f1613_g.lkr, lkr/12lf1612_g.lkr,
    lkr/16lf1613_g.lkr:
    New devices based on the mplabx-2.0.

  * header/p12f1571.inc, header/p12f1572.inc, header/p12lf1571.inc,
    header/p12lf1572.inc, header/p16f1703.inc, header/p16f1707.inc,
    header/p16f1713.inc, header/p16f1716.inc, header/p16f1782.inc,
    header/p16f1783.inc, header/p16f1784.inc, header/p16f1786.inc,
    header/p16f1787.inc, header/p16f1788.inc, header/p16f1789.inc,
    header/p16lf1703.inc, header/p16lf1707.inc, header/p16lf1713.inc,
    header/p16lf1716.inc, header/p16lf1782.inc, header/p16lf1783.inc,
    header/p16lf1784.inc, header/p16lf1786.inc, header/p16lf1787.inc,
    header/p16lf1788.inc, header/p16lf1789.inc, lkr/*_g.lkr:
    Updated based on the mplabx-2.0.

2014-01-22  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/p12f1822.inc, header/p12f1840.inc, header/p12lf1822.inc,
    header/p12lf1840.inc, header/p12lf1840t39a.inc, header/p12lf1840t48a.inc:
    Added the missing CONFIG2 Options: _DEBUG_ON, _DEBUG_OFF
    "#62 Missing CONFIG2 DEBUG file definition in some pic12f inc files"
    thanks to Ludrao.

2013-11-09  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p16f1703.inc, header/p16f1705.inc, header/p16f1707.inc,
    header/p16f1709.inc, header/p16lf1703.inc, header/p16lf1705.inc,
    header/p16lf1707.inc, header/p16lf1709.inc,
    libgputils/gpcfg-table.c, libgputils/gpprocessor.c,
    lkr/16f1703_g.lkr, lkr/16f1705_g.lkr, lkr/16f1707_g.lkr,
    lkr/16f1709_g.lkr, lkr/16lf1703_g.lkr, lkr/16lf1705_g.lkr,
    lkr/16lf1707_g.lkr, lkr/16lf1709_g.lkr, 
    lkr/Makefile.am, lkr/Makefile.in:
    New devices based on the mplabx-1.95.

  * header/p12f1571.inc, header/p12f1572.inc, header/p12f1822.inc,
    header/p12lf1571.inc, header/p12lf1572.inc, header/p16f1784.inc,
    header/p16lf1784.inc,
    lkr/12f1572_g.lkr, lkr/12lf1572_g.lkr, lkr/16f1509_g.lkr,
    lkr/16f1783_g.lkr, lkr/16f1936_g.lkr, lkr/16f1937_g.lkr,
    lkr/16lf1509_g.lkr, lkr/16lf1783_g.lkr, lkr/16lf1906_g.lkr,
    lkr/16lf1907_g.lkr, lkr/16lf1936_g.lkr, lkr/16lf1937_g.lkr:
    Updated based on the mplabx-1.95.

  * AUTHORS:
    Corrected my email.

2013-10-12  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpmemory.c: fixed b_memory_get()

2013-10-11  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpmemory.c: fixed bug
    #269 18F26j50, gplink crash on WPDIS

2013-09-25  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/scan.l, gpasm/ppscan.l: fixed bug
    #270 Gputils doesn't compile with newer versions of bison and flex

2013-09-08  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p12f1571.inc, header/p12f1572.inc, header/p12lf1571.inc,
    header/p12lf1572.inc, header/p16f1713.inc, header/p16f1716.inc,
    header/p16f1829lin.inc, header/p16lf1713.inc, header/p16lf1716.inc,
    libgputils/gpcfg-table.c, libgputils/gpprocessor.c,
    lkr/12f1571_g.lkr, lkr/12f1572_g.lkr, lkr/12lf1571_g.lkr,
    lkr/12lf1572_g.lkr, lkr/16f1713_g.lkr, lkr/16f1716_g.lkr,
    lkr/16f1829lin_g.lkr, lkr/16lf1713_g.lkr, lkr/16lf1716_g.lkr,
    lkr/Makefile.am, lkr/Makefile.in:
    New devices based on the mplabx-1.9.

2013-09-08  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/*.inc:
    Updated based on the mplabx-1.9.

2013-08-09  Borut Razem  <borutr@users.sourceforge.net>
  * config.guess, config.sub: updated from git head
  * depcomp: updated
  * README: updated MinGW build info

2013-08-05  Borut Razem  <borutr@users.sourceforge.net>
  * gplink/gplink.c, libgputils/gpcofflink.c, libgputils/gpcofflink.h,
    gplink/gplink.h.in, doc/gputils.lyx, man/gplink.1.in,
    man/fr/gplink.1.in: implemented --mplink-compatble in gplink
  * gplink/testsuite/gplink.project/asmfiles/op14e.asm,
    gplink/testsuite/gplink.project/lkrfiles/op14e.lkr,

trunk/gputils/gputils/gpvo.c
/* GNU PIC view object
   Copyright (C) 2001, 2002, 2003, 2004, 2005
   Craig Franklin

    Copyright (C) 2016 Molnar Karoly <molnarkaroly@users.sf.net>

This file is part of gputils.

gputils is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

gputils is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with gputils; see the file COPYING.  If not, write to
the Free Software Foundation, 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.  */

#include "stdhdr.h"

#include "libgputils.h"
#include "gpvo.h"

struct gpvo_state state;

enum {
  OPT_STRICT_OPTIONS = 0x100
};

#define GET_OPTIONS "bcfhnstvx:y"

/* Used: himpsv */
static struct option longopts[] =
{
  { "binary",         no_argument,       NULL, 'b' },
  { "mnemonics",      no_argument,       NULL, 'c' },
  { "file",           no_argument,       NULL, 'f' },
  { "help",           no_argument,       NULL, 'h' },
  { "no-names",       no_argument,       NULL, 'n' },
  { "section",        no_argument,       NULL, 's' },
  { "strict-options", no_argument,       NULL, OPT_STRICT_OPTIONS },
  { "symbol",         no_argument,       NULL, 't' },
  { "version",        no_argument,       NULL, 'v' },
  { "export",         required_argument, NULL, 'x' },
  { "extended",       no_argument,       NULL, 'y' },
  { NULL,             no_argument,       NULL, '\0'}
};

/*------------------------------------------------------------------------------------------------*/

static void
_show_usage(void)
{
  printf("Usage: gpvo [options] file\n");
  printf("Options: [defaults in brackets after descriptions]\n");
  printf("  -b, --binary            Print binary data.\n");
  printf("  -c, --mnemonics         Decode special mnemonics.\n");
  printf("  -f, --file              File header.\n");
  printf("  -h, --help              Show this usage message.\n");
  printf("  -n, --no-names          Suppress filenames.\n");
  printf("  -s, --section           Section data.\n");
  printf("      --strict-options    If this is set, then an option may not be parameter\n"
         "                            of an another option. For example: -x --symbol\n");
  printf("  -t  --symbol            Symbol table.\n");
  printf("  -v, --version           Show version.\n");
  printf("  -x FILE, --export FILE  Export symbols to include file.\n");
  printf("  -y, --extended          Enable 18xx extended mode.\n\n");
  printf("Report bugs to:\n");
  printf("%s\n", PACKAGE_BUGREPORT);
  exit(0);
}

/*------------------------------------------------------------------------------------------------*/

static void
_print_header(const gp_object_t *object)
{
  static struct magic_s {
    uint16_t    magic_num;
    const char *magic_name;
  } magic[] = {
    { 0x1234, "MICROCHIP_MAGIC_v1" },
    { 0x1240, "MICROCHIP_MAGIC_v2" }
  };

  time_t      time;
  char       *time_str;
  const char *processor_name;
  int         i;

  time     = (time_t)object->time;
  time_str = ctime(&time);
  processor_name = gp_processor_name(object->processor, 2);

  /* strip the newline from time */
  time_str[strlen(time_str) - 1] = '\0';

  printf("COFF File and Optional Headers\n");

  for (i = 0; i < ARRAY_SIZE(magic); ++i) {
    if (magic[i].magic_num == object->version) {
      break;
    }
  }
  printf("COFF version         %#x: %s\n", object->version,
         (i < ARRAY_SIZE(magic)) ? magic[i].magic_name : "unknown");
  printf("Processor Type       %s\n",  processor_name);
  printf("Time Stamp           %s\n",  time_str);
  printf("Number of Sections   %zu\n", object->section_list.num_nodes);
  printf("Number of Symbols    %u\n",  object->num_symbols);
  printf("Characteristics      %#x\n", object->flags);

  if (object->flags & F_RELFLG) {
    printf("  Relocation info has been stripped.\n");
  }

  if (object->flags & F_EXEC) {
    printf("  File is executable.\n");
  }

  if (object->flags & F_LINENO) {
    printf("  Line numbers have been stripped.\n");
  }

  if (object->flags & F_ABSOLUTE) {
    printf("  The MPASM assembler object file is from absolute assembly code.\n");
  }

  if (object->flags & L_SYMS) {
    printf("  Local symbols have been stripped.\n");
  }

  if (object->flags & F_EXTENDED18) {
    printf("  The COFF file produced utilizing the Extended mode.\n");
  }

  if (object->flags & F_GENERIC) {
    printf("  Processor independent file for a core.\n");
  }

  printf("\n");
}

/*------------------------------------------------------------------------------------------------*/

static const char *
_format_reloc_type(uint16_t type, char *buffer, size_t sizeof_buffer)
{
  snprintf(buffer, sizeof_buffer, "0x%04x %-20s", type, gp_coffgen_reloc_type_to_str(type));
  return buffer;
}

/*------------------------------------------------------------------------------------------------*/

static void
_print_relocation_list(proc_class_t class, const gp_reloc_t *relocation)
{
  char        buffer[32];
  int         addr_digits;
  const char *column_gap;

  addr_digits = class->addr_digits;
  column_gap  = (addr_digits > 4) ? "" : "  ";

  printf("Relocations Table\n"
         "Address     Offset    Type                        Symbol\n");

  while (relocation != NULL) {
    printf("0x%0*x%s    0x%0*x%s  %-25s %-s\n",
           addr_digits, gp_processor_byte_to_org(class, relocation->address), column_gap,
           addr_digits, relocation->offset, column_gap,
           _format_reloc_type(relocation->type, buffer, sizeof(buffer)),
           relocation->symbol->name);

    relocation = relocation->next;
  }

  printf("\n");
}

/*------------------------------------------------------------------------------------------------*/

static void
_print_linenum_list(proc_class_t class, const gp_linenum_t *linenumber)
{
  const char *filename;
  int         addr_digits;
  const char *column_gap;

  addr_digits = class->addr_digits;
  column_gap  = (addr_digits > 4) ? "  " : "    ";

  printf("Line Number Table\n"
         "Line      Address     Symbol\n");

  while (linenumber != NULL) {
    if (state.suppress_names) {
      filename = linenumber->symbol->name;
    }
    else {
      filename = linenumber->symbol->aux_list.first->_aux_symbol._aux_file.filename;
    }

    printf("%-8i  0x%0*x%s  %s\n",
           linenumber->line_number,
           addr_digits, gp_processor_byte_to_org(class, linenumber->address), column_gap,
           filename);

    linenumber = linenumber->next;
  }

  printf("\n");
}

/*------------------------------------------------------------------------------------------------*/

static void
_print_data(proc_class_t class, pic_processor_t processor, const gp_section_t *section)
{
  char         buffer[BUFSIZ];
  unsigned int address;
  int          num_words;
  int          addr_digits;
  unsigned int bsr_boundary;
  uint16_t     word;
  uint8_t      byte;

  address      = section->address;
  bsr_boundary = gp_processor_bsr_boundary(processor);
  addr_digits  = class->addr_digits;

  buffer[0] = '\0';

  printf("Data\n");
  while (true) {
    if ((section->flags & STYP_TEXT) && (class->find_insn != NULL)) {
      if (!class->i_memory_get(section->data, address, &word, NULL, NULL)) {
        break;
      }

      num_words = gp_disassemble(section->data, address, class, bsr_boundary, 0,
                                 GPDIS_SHOW_ALL_BRANCH, buffer, sizeof(buffer), 0);
      printf("%0*x:  %04x  %s\n",
             addr_digits, gp_processor_byte_to_org(class, address), word, buffer);

      if (num_words != 1) {
        if (!class->i_memory_get(section->data, address + WORD_SIZE, &word, NULL, NULL)) {
          break;
        }

        printf("%0*x:  %04x\n",
               addr_digits, gp_processor_byte_to_org(class, address + WORD_SIZE), word);
      }

      address += num_words * WORD_SIZE;
    }
    else if ((section->flags & STYP_DATA_ROM) || (class == PROC_CLASS_EEPROM16)) {
      if (class->i_memory_get(section->data, address, &word, NULL, NULL)) {
        printf("%0*x:  %04x\n", addr_digits, gp_processor_byte_to_org(class, address), word);
        address += WORD_SIZE;
      }
      else {
        if (b_memory_get(section->data, address, &byte, NULL, NULL)) {
          printf("%0*x:  %02x\n", addr_digits, gp_processor_byte_to_org(class, address), byte);
        }

        break;
      }
    }
    else {
      /* STYP_DATA or STYP_ACTREC, or EEPROM8 */
      if (!b_memory_get(section->data, address, &byte, NULL, NULL)) {
        break;
      }

      printf("%0*x:  %02x\n", addr_digits, address, byte);
      ++address;
    }
  }
  printf("\n");
}

/*------------------------------------------------------------------------------------------------*/

static void
_print_sec_header(proc_class_t class, const gp_section_t *section)
{
  unsigned int org_to_byte_shift;

  org_to_byte_shift = (section->flags & STYP_ROM_AREA) ? class->org_to_byte_shift : 0;

  printf("Section Header\n");
  printf("Name                    %s\n",  section->name);
  printf("Physical address        %#x\n", gp_byte_to_org(org_to_byte_shift, section->address));
  printf("Virtual address         %#x\n", gp_byte_to_org(org_to_byte_shift, section->virtual_address));
  printf("Size of Section         %u\n",  section->size);
  printf("Number of Relocations   %zu\n", section->relocation_list.num_nodes);
  printf("Number of Line Numbers  %zu\n", section->line_number_list.num_nodes);
  printf("Flags                   %#x\n", section->flags);

  if (section->flags & STYP_TEXT) {
    printf("  Executable code.\n");
  }

  if (section->flags & STYP_DATA) {
    printf("  Initialized data.\n");
  }

  if (section->flags & STYP_BSS) {
    printf("  Uninitialized data.\n");
  }

  if (section->flags & STYP_DATA_ROM) {
    printf("  Initialized data for ROM.\n");
  }

  if (section->flags & STYP_ABS) {
    printf("  Absolute.\n");
  }

  if (section->flags & STYP_SHARED) {
    printf("  Shared across banks.\n");
  }

  if (section->flags & STYP_OVERLAY) {
    printf("  Overlaid with other sections from different objects modules.\n");
  }

  if (section->flags & STYP_ACCESS) {
    printf("  Available using access bit.\n");
  }

  if (section->flags & STYP_ACTREC) {
    printf("  Contains the activation record for a function.\n");
  }

  printf("\n");
}

/*------------------------------------------------------------------------------------------------*/

static void
_print_sec_list(const gp_object_t *object)
{
  const gp_section_t *section;

  section = object->section_list.first;
  while (section != NULL) {
    _print_sec_header(object->class, section);

    if ((section->size > 0) && (section->data_ptr > 0)) {
      _print_data(object->class, object->processor, section);
    }

    if (section->relocation_list.num_nodes > 0) {
      _print_relocation_list(object->class, section->relocation_list.first);
    }

    if (section->line_number_list.num_nodes > 0) {
      _print_linenum_list(object->class, section->line_number_list.first);
    }

    section = section->next;
  }
}

/*------------------------------------------------------------------------------------------------*/

static void
_coff_type(unsigned int type, char *buffer, size_t sizeof_buffer)
{
  const char *str;

  switch (type) {
    case T_NULL:   str = "NULL";                break; /* null */
    case T_VOID:   str = "void";                break; /* void */
    case T_CHAR:   str = "char";                break; /* character */
    case T_SHORT:  str = "short";               break; /* short integer */
    case T_INT:    str = "int";                 break; /* integer */
    case T_LONG:   str = "long int";            break; /* long integer */
    case T_FLOAT:  str = "float";               break; /* floating point */
    case T_DOUBLE: str = "double";              break; /* double length floating point */
    case T_STRUCT: str = "struct";              break; /* structure */
    case T_UNION:  str = "union";               break; /* union */
    case T_ENUM:   str = "enum";                break; /* enumeration */
    case T_MOE:    str = "enum";                break; /* member of enumeration */
    case T_UCHAR:  str = "unsigned char";       break; /* unsigned character */
    case T_USHORT: str = "unsigned short";      break; /* unsigned short */
    case T_UINT:   str = "unsigned int";        break; /* unsigned integer */
    case T_ULONG:  str = "unsigned long";       break; /* unsigned long */
    case T_LNGDBL: str = "long double";         break; /* long double floating point */
    case T_SLONG:  str = "short long";          break; /* short long */
    case T_USLONG: str = "unsigned short long"; break; /* unsigned short long */
    default:       str = NULL;                  break;
  }

  if (str != NULL) {
    snprintf(buffer, sizeof_buffer, str);
  }
  else {
    snprintf(buffer, sizeof_buffer, "unknown(%u)", type);
  }
}

/*------------------------------------------------------------------------------------------------*/

static const char *
_format_sym_type(unsigned int type, char *buffer, size_t sizeof_buffer)
{
  const char *str;

  str = gp_coffgen_symbol_type_to_str(type);
  if (str != NULL) {
    return str;
  }

  snprintf(buffer, sizeof_buffer, "%u", type);
  return buffer;
}

/*------------------------------------------------------------------------------------------------*/

static const char *
_format_sym_derived_type(unsigned int type, char *buffer, size_t sizeof_buffer)
{
  const char *str;

  str = gp_coffgen_symbol_derived_type_to_str(type);
  if (str != NULL) {
    return str;
  }

  snprintf(buffer, sizeof_buffer, "%u", type);
  return buffer;
}

/*------------------------------------------------------------------------------------------------*/

static const char *
_format_sym_class(unsigned int cls, char *buffer, size_t sizeof_buffer)
{
  const char *str;

  str = gp_coffgen_symbol_class_to_str(cls);
  if (str != NULL) {
    return str;
  }

  snprintf(buffer, sizeof_buffer, "%u", cls);
  return buffer;
}

/*------------------------------------------------------------------------------------------------*/

#define AUX_INDENT              "      "

static void
_print_sym_table(const gp_object_t *object)
{
  static char  buf[64];

  gp_symbol_t *symbol;
  gp_symbol_t *callee;
  gp_aux_t    *aux;
  const char  *section;
  int          c;
  int          i;
  int          idx = 1;
  char         buffer_type[8];
  char         buffer_derived_type[8];
  char         buffer_class[8];

  printf("Symbol Table\n");
  printf("Idx  Name                     Section          Value      Type     DT           Class     NumAux\n");

  symbol = object->symbol_list.first;
  while (symbol != NULL) {
    if (symbol->section_number == N_DEBUG) {
      section = "DEBUG";
    }
    else if (symbol->section_number == N_ABS) {
      section = "ABSOLUTE";
    }
    else if (symbol->section_number == N_UNDEF) {
      section = "UNDEFINED";
    }
    else {
      if (symbol->section != NULL) {
        section = symbol->section->name;
      }
      else {
        snprintf(buf, sizeof(buf), "Bad num.: %u", symbol->section_number);
        section = buf;
      }
    }

    printf("%04d %-24s %-16s %#-10lx %-8s %-12s %-9s %-4zu\n",
           idx,
           symbol->name,
           section,
           symbol->value,
           _format_sym_type(symbol->type, buffer_type, sizeof(buffer_type)),
           _format_sym_derived_type(symbol->derived_type, buffer_derived_type, sizeof(buffer_derived_type)),
           _format_sym_class(symbol->class, buffer_class, sizeof(buffer_class)),
           symbol->aux_list.num_nodes);

    aux = symbol->aux_list.first;
    while (aux != NULL) {
      switch (aux->type) {
        case AUX_DIRECT:
          printf(AUX_INDENT "command = \"%c\"\n", aux->_aux_symbol._aux_direct.command);
          printf(AUX_INDENT "string  = \"%s\"\n", aux->_aux_symbol._aux_direct.string);
          break;

        case AUX_FILE: {
          if (!state.suppress_names) {
            printf(AUX_INDENT "file = %s\n", aux->_aux_symbol._aux_file.filename);
          }
          printf(AUX_INDENT "line included = %u\n", aux->_aux_symbol._aux_file.line_number);
          printf(AUX_INDENT "flags         = %x\n", aux->_aux_symbol._aux_file.flags);
          break;
        }

        case AUX_IDENT:
          printf(AUX_INDENT "string = \"%s\"\n", aux->_aux_symbol._aux_ident.string);
          break;

        case AUX_SCN:
          printf(AUX_INDENT "length                 = %u\n", aux->_aux_symbol._aux_scn.length);
          printf(AUX_INDENT "number of relocations  = %u\n", aux->_aux_symbol._aux_scn.nreloc);
          printf(AUX_INDENT "number of line numbers = %u\n", aux->_aux_symbol._aux_scn.nlineno);
          break;

        case AUX_FCN_CALLS: {
          callee = aux->_aux_symbol._aux_fcn_calls.callee;
          printf(AUX_INDENT "callee       = %s\n", (callee != NULL) ? callee->name : "higher order");
          printf(AUX_INDENT "is_interrupt = %u\n", aux->_aux_symbol._aux_fcn_calls.is_interrupt);
          break;
        }

        default: {
          printf("%u  ", aux->type);
          for (i = 0; i < object->symbol_size; i++) {
            printf("%02x", aux->_aux_symbol.data[i] & 0xFF);

            if (i & 1) {
              putchar(' ');
            }
          }
          for (i = 0; i < object->symbol_size; i++) {
            c = aux->_aux_symbol.data[i] & 0xFF;
            putchar((isprint(c)) ? c : '.');
          }
          putchar('\n');
        }
      } /* switch (aux->type) */

      aux = aux->next;
      ++idx;
    }

    symbol = symbol->next;
    ++idx;
  }

  printf("\n");
}

/*------------------------------------------------------------------------------------------------*/

static void
_export_sym_table(gp_object_t *object)
{
  gp_symbol_t *symbol;
  char         buffer[BUFSIZ];

  symbol = object->symbol_list.first;
  while (symbol != NULL) {
    if ((state.export.enabled) && (symbol->class == C_EXT) && (symbol->section_number > N_UNDEF)) {
      _coff_type(symbol->type, buffer, sizeof(buffer));
      fprintf(state.export.f, "  extern %s ; %s\n", symbol->name, buffer);
    }

    symbol = symbol->next;
  }
}

/*------------------------------------------------------------------------------------------------*/

#define BIN_DATA_LINE_SIZE    16

static void
_print_binary(const uint8_t *data, long file_size)
{
  long         i;
  long         j;
  unsigned int memory;
  int          c;

  printf("\nObject file size = %li bytes\n", file_size);

  printf("\nBinary object file contents:");
  for (i = 0; i < file_size; i += BIN_DATA_LINE_SIZE) {
    printf("\n%06lx", i);

    for (j = 0; j < BIN_DATA_LINE_SIZE; j += 2) {
      memory = (data[i + j] << 8) | data[i + j + 1];

      if ((i + j) >= file_size) {
        printf("     ");
      }
      else {
        printf(" %04x", memory);
      }
    }

    printf(" ");

    for (j = 0; j < BIN_DATA_LINE_SIZE; j += 2) {
      if ((i + j) < file_size) {
        c = data[i + j];
        putchar((isprint(c)) ? c : '.');

        c = data[i + j + 1];
        putchar((isprint(c)) ? c : '.');
      }
    }
  }

  printf("\n\n");
}

/*------------------------------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
  int         option_index;
  int         c;
  const char *command;
  gp_boolean  strict_options = false;
  gp_boolean  usage          = false;
  char        buffer[BUFSIZ];

  gp_init();

  /* initalize */
  state.dump_flags     = 0;
  state.suppress_names = false;
  state.export.enabled = false;

  /* Scan through the options for the --strict-options flag. */
  while ((c = getopt_long(argc, argv, GET_OPTIONS, longopts, NULL)) != EOF) {
    if (c == OPT_STRICT_OPTIONS) {
      strict_options = true;
      break;
    }
  }

  /* Restores the getopt_long index. */
  optind = 1;
  while (true) {
    /* This is necessary for the gp_exit_is_excluded_arg() function. */
    option_index = -1;
    command = argv[optind];
    if ((c = getopt_long(argc, argv, GET_OPTIONS, longopts, &option_index)) == EOF) {
      break;
    }

    if (strict_options) {
      gp_exit_if_arg_an_option(longopts, ARRAY_SIZE(longopts), option_index, optarg, c, command);
    }

    switch (c) {
    case '?':
    case 'h':
      usage = true;
      break;

    case 'b':
      state.dump_flags |= PRINT_BINARY;
      break;

    case 'c':
      gp_decode_mnemonics = true;
      break;

    case 'f':
      state.dump_flags |= PRINT_HEADER;
      break;

    case 'n':
      state.suppress_names = true;
      break;

    case 's':
      state.dump_flags |= PRINT_SECTIONS;
      break;

    case 't':
      state.dump_flags |= PRINT_SYMTBL;
      break;

    case 'y':
      gp_decode_extended = true;
      break;

    case 'x':
      state.export.enabled = true;
      state.export.filename = optarg;
      break;

    case 'v':
      fprintf(stderr, "%s\n", GPVO_VERSION_STRING);
      exit(0);

    case OPT_STRICT_OPTIONS:
      /* do nothing */
      break;
    }

    if (usage)
      break;
  }

  if ((optind + 1) == argc) {
    state.filename = argv[optind];
  }
  else {
    usage = true;
  }

  if (usage) {
    _show_usage();
  }

  if (!state.dump_flags) {
    /* no command line flags were set so print everything */
    state.dump_flags = PRINT_HEADER | PRINT_SECTIONS | PRINT_SYMTBL;
  }

  if ((gp_identify_coff_file(state.filename) != GP_COFF_OBJECT_V2) &&
      (gp_identify_coff_file(state.filename) != GP_COFF_OBJECT)) {
    gp_error("\"%s\" is not a valid object file", state.filename);
    exit(1);
  }

  state.object = gp_read_coff(state.filename);
  state.file   = gp_read_file(state.filename);

  if (state.export.enabled) {
    state.export.f = fopen(state.export.filename, "w");

    if (state.export.f == NULL) {
      perror(state.export.filename);
      exit(1);
    }

    gp_date_string(buffer, sizeof(buffer));

    fprintf(state.export.f, "; %s\n", state.export.filename);
    fprintf(state.export.f, "; generated by %s on %s\n", GPVO_VERSION_STRING, buffer);
    fprintf(state.export.f, "; from %s\n\n", state.filename);

    _export_sym_table(state.object);

    fclose(state.export.f);

    /* suppress normal output */
    state.dump_flags = 0;
  }

  if (state.dump_flags & PRINT_HEADER) {
    _print_header(state.object);
  }

  if (state.dump_flags & PRINT_SECTIONS) {
    _print_sec_list(state.object);
  }

  if (state.dump_flags & PRINT_SYMTBL) {
    _print_sym_table(state.object);
  }

  if (state.dump_flags & PRINT_BINARY) {
    _print_binary(state.file->file, state.file->size);
  }

  return EXIT_SUCCESS;
}
